retryRpc package to implement retryable RPCs with exactly once semantics

bschatz-swift opened the pull request:
api.go - API for package with comments to show algorithm and open issues
api_test.go - shows how the API would be used by the caller

Commit:
Prototype bypass of SwiftProxy for PFSAgent JSON RPC directly to ProxyFS

Commit:
Update PFSAgent SwiftProxy Bypass to properly frame JSON Replies

bschatz-swift committed:
First pass at retryrpc API and sample usage

bschatz-swift committed:
Hash out more of the API

bschatz-swift committed:
Add algorithm to APIs and update Send() method to reflect how called

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+)
+
+// Request is the structure sent
+type Request struct {
```
So by this type (Request) and the corresponding type Response, we are jettisoning actual JSON RPC, correct? Sure, we've decided to marshal via JSON... but we no longer need anything like the form that JSON RPC defines namely:
```
type jrpcRequestStruct struct {
	JSONrpc string         `json:"jsonrpc"`
	Method  string         `json:"method"`
	ID      uint64         `json:"id"`
	Params  [1]interface{} `json:"params"`
}

type jrpcResponseWithErrorStruct struct {
	ID     uint64      `json:"id"`
	Error  string      `json:"error"`
	Result interface{} `json:"result"`
}
```
Correct?

bschatz-swift replied:
Request.JReq will contain JSON with these fields:

    type jsonBypassRequest struct {
    	    Method  string         `json:"method"`
	    Params  [1]interface{} `json:"params"`
    }

and the Response.Res will contain JSON with these fields:

    typeBypassResponse struct {
             Result interface{}  `json:"result"`
    }

I will not be using jrpcRequestStruct or jrpcResponseWithErrorStruct.

Did that answer your question?

I will update the PR with these responses to include the marshaling and unmarshaling.

edmc-ss replied:
Yeah, as I mentioned verbally, I doubt we need the whole JSON-RPC doc in there...

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+	completedTTL      time.Duration // How long a completed request stays on queue
+	serviceMap        map[string]*func(request interface{}, response interface{})
+	connections       *list.List                   // List of outstanding connections
+	pendingRequests   map[string]*Request          // Request which have not been completed yet.
```
For maps, I like to be told at least what the Key is... I can (usually) see what the Value is :-).

As for the Key being a `string`, I'm actually surprised at this. Perhaps you do something with RequestID and MountID to be the string that is the Key here? There are more direct ways to do that.

For instance, you can look at type `readCacheKeyStruct` in inode/config.go:
```
type readCacheKeyStruct struct {
	volumeName       string
	logSegmentNumber uint64
	cacheLineTag     uint64 // LogSegment offset / readCacheLineSize
}
```
It is used as the Key in the readCache field of the volumeGroupStruct listed just below.
In this case, since it's a named Key type, I didn't bother commenting what it is :-).

Anyway, this is just a cute thing you can do with Go maps I thought more appropriate here.

bschatz-swift replied:
Good catch.   The key for pendingRequests will be a combination of mountID and client request ID.   I added this to the code:

// RequestKey uniquely identifies a client request from a client
type RequestKey struct {
	uniqueID string
	clientRequestID uint64
}

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+	sync.Mutex
+	completedTTL      time.Duration // How long a completed request stays on queue
+	serviceMap        map[string]*func(request interface{}, response interface{})
+	connections       *list.List                   // List of outstanding connections
```
As per my comment just following, just saying *list.List fails to reference what the Values (Elements) actually are... a comment would be more helpful if it referenced a struct corresponding to the Connections that are on the list. You certainly need such a structure if only to wrap a net.Conn and the list.Element, right?

bschatz-swift replied:
Not sure yet how I will manage the connections.   I think I will put it in the request that I queue on pendingRequest.   I will work on that.

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+	// 1. goroutine which accepts new connection and creates goroutine W
+	// 2. goroutine W adds request to server.pendingRequest, unmarshals
+	//    request, processes RPC, puts request on server.completedRequest
+	//    and under same lock removes from server.pendingRequest.
```
I would suggest the "processes RPC" step etc... be done in a per-request Goroutine itself, right?

bschatz-swift replied:
Yes - updated comment

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+
+// Client methods
+
+// Client tracking structure
```
So I gather this `retryrpc` package will be used by both package jrpcfs (or its replacement/peer) and program pfsagentd, correct? Sounds great!

bschatz-swift replied:
yes - that is correct

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+	// tick at mount and increment from there?
+	// Handle reset of time?
+	myUniqueID         uint64              // Unique ID across all clients
+	subsetID           uint64              // Subset of unique ID (mount ID)
```
Re myUniqueID & subsetID... I don't think either make too much sense.
Don't we just want to have a []byte that holds whatever the client will be passing in the MountID field mentioned above? I was mostly confused why there are two such fields here...

bschatz-swift replied:
Good - I wanted to talk about this.

We said we wanted an opaque ID since we may break this off into a separate repo.  (That is why I did not name it mountID)

I guess in our case pfsagent will have a different mountID per mounted file system.   We do not have to worry about one pfsagent handling multiple mount points.   If we did, then we would have to have a unique ID for the client and a unique ID for the mount points it was using.

Therefore, myUniqueID is the only field we need.

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+// Client methods
+
+// Client tracking structure
+type Client struct {
```
PFSAgent will only have one of these, correct?
I mean it does a single mount... and opens up a single connection.
I suppose other clients may want to manage multiple... presumably, they'd each have a unique connection to the server, right? We should probably make that clear by embedding in this structure things like the connection... Presumably, there'd also be a "shuttingDown" bool that is set when shutting down :-) and the Close() method sets before closing the network connection... thus the goroutine reading the connection would see the connection go down and check the "shuttingDown" bool to know that this was intentional (rather than re-connecting and re-transmitting each request).

bschatz-swift replied:
Great!   That is what I wanted to clean up.    Yes - PFSAgent will only have one Client struct now that we are sure there is one PFSAgent per mount point.

I agree with the fields for shutting down, etc.   I am looking for feedback on the overall approach.

edmc-ss commented on retryrpc/api.go:
```diff
@@ -0,0 +1,134 @@
+	//       the operation idempotent.   Assume client retries until
+	//       server comes back up?   Wait for failover to a peer?
+	//       Assume using VIP on proxyfs node?
+	// 4. Should we block forever? How kill?
```
I think you need the Client.outstandingRequest map (hmm... do we need to re-transmit in time order and, hence, also have to maintain a time-ordered list of them?)... and to have a separate "Sender" goroutine. Similarly, there should be a "Receiver" goroutine that receives (out-of-order) responses looking them up in Client.outstandingRequest to know how to signal the Send() func that is blocked waiting on the completion, right? Thus, in Request, there should be something like a sync.WaitGroup and, perhaps, an error as well in case the RPC was "unsendable".

bschatz-swift committed:
Incorporate review feedback

bschatz-swift committed:
Add Register() functionality

bschatz-swift committed:
Start client/server work

bschatz-swift committed:
client to server messaging

bschatz-swift committed:
Add unmarshaling

bschatz-swift committed:
Full RpcPing clnt, svr and back to client

bschatz-swift committed:
Cleanup the locking and close holes

bschatz-swift committed:
Rough out the I/O context algorithm

bschatz-swift committed:
Client side goroutine pipeline

Needs restransmit if I/O error but as multi-threaded as I can make it at
this point.

bschatz-swift commented:
The latest code incorporates these comments as needed so this branch can be deleted
